soc/xtensa/intel_adsp/common/include/soc.h

```diff
@@ -72,9 +66,62 @@
 #define DSP_WCT_CS_TA(x)			BIT(x)
 #define DSP_WCT_CS_TT(x)			BIT(4 + x)
 
+/* Attribute macros to place code and data into IMR memory */
+#define __imr __in_section_unique(imr)
+#define __imrdata __in_section_unique(imrdata)
+
 extern void z_soc_irq_enable(uint32_t irq);
 extern void z_soc_irq_disable(uint32_t irq);
 extern int z_soc_irq_is_enabled(unsigned int irq);
 
+/* Legacy SOC-level API still used in a few drivers */
+#define SOC_DCACHE_FLUSH(addr, size)		\
+	z_xtensa_cache_flush((addr), (size))
+#define SOC_DCACHE_INVALIDATE(addr, size)	\
+	z_xtensa_cache_inv((addr), (size))
+
+/**
+ * @brief Return uncached pointer to a RAM address
+ *
+ * The Intel ADSP architecture maps all addressable RAM (of all types)
+ * twice, in two different 512MB segments regions whose L1 cache
+ * settings can be controlled independently.  So for any given
+ * pointer, it is possible to convert it to and from a cached version.
+ *
+ * This function takes a pointer to any addressible object (either in
+ * cacheable memory or not) and returns a pointer that can be used to
+ * refer to the same memory while bypassing the L1 data cache.  Data
+ * in the L1 cache will not be inspected nor modified by the access.
+ *
+ * @see z_soc_cached_ptr()
+ *
+ * @param p A pointer to a valid C object
+ * @return A pointer to the same object bypassing the L1 dcache
+ */
+static inline void *z_soc_uncached_ptr(void *p)
+{
+	return ((void *)(((size_t)p) & ~0x20000000));
```

Comment:
I maybe misinterpretting here, but the address has a bit that signifies cacheable or uncacheable as I read this, perhaps defining the bit with BIT() and using that define here would be a nice way to make this a little more readable.

Reply:
(So... this is another comment that actually belongs in the earlier PR, heh.  But no reason not to discuss it here.)

It's actually an offset, not a bit.  The bit math is an optimization.

Xtensa's "region protection option" partitions memory into 8 regions, each of which can be provided different protection settings, including cache control.  The hardware designers mapped the RAM blocks twice, once in 0x80000000 and again at 0xa000000.  And we exploit that so we can put some symbols or regions into one "space" or the other and control their cacheability via what pointer we use to access them (with great care, obviously!).

As it happens those are adjacent so we can effect the offset by toggling bit 0x20000000.  But that's an artifact of the hardware mapping decision.

Other hardware (ahem, watch this space) might make different decisions and (sigh) use a non-power-of-two offset.  And it needs a different abstraction still than you see here.

Reply:
That completely makes sense now, I read the comment here but didn't quite grok it, but the comment does describe just this already I believe. Thanks!